### PR TITLE
Add focusRef to TouchableWin32

### PR DESCRIPTION
**Pull request by tkamireh**
Allows imperative focusing of TouchableWin32s

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5108)

### CODE REVIEW

File: packages/react-native-win32/src/Libraries/Components/Touchable/Tests/TouchableWin32Test.tsx [hunk 1]
```diff
@@ -424,23 +429,65 @@ class TouchableHighlightExample extends React.Component<{}, IExampleState> {
   };
 }
 
+const TouchableFocusExample = () => {
+  const [focused, setFocused] = React.useState(false);
+  let focusableRef = React.useRef<ViewWin32>(null);
```
Comment (PPatBoyd) [inline packages/react-native-win32/src/Libraries/Components/Touchable/Tests/TouchableWin32Test.tsx:434]: you need to use setAndForwardRef a la ViewWin32 (doesn't need to be a function component), this breaks using TouchableWin32 as a refobject

File: packages/react-native-win32/src/Libraries/Components/Touchable/TouchableWin32.tsx [hunk 2]
```diff
@@ -257,6 +263,10 @@ export class TouchableWin32 extends React.Component<ITouchableWin32Props, IInter
     );
   }
 
+  public focus = () => {
```
Comment (tkamireh) [inline packages/react-native-win32/src/Libraries/Components/Touchable/TouchableWin32.tsx:266]: @PPatBoyd, this is one of the approaches I was talking about offline, effectively for class components we can wrap everything in this way. Outer components hold RefObject<TouchableWin32>s that can imperatively call focus on. I think there's a similar thing we can do with function components using imperativeHandles.

File: packages/react-native-win32/src/Libraries/Components/Touchable/TouchableWin32.tsx [hunk 3]
```diff
@@ -257,6 +263,10 @@ export class TouchableWin32 extends React.Component<ITouchableWin32Props, IInter
...
+    this._internalRef.current.focus();
```
Comment (Dewsk) [inline packages/react-native-win32/src/Libraries/Components/Touchable/TouchableWin32.tsx:267]: Do you need to provide any nullish checking here?

**Comment by TheSavior** [comment]
In core when you attach a ref to TouchableHighlight, we [use forwardRef](https://github.com/facebook/react-native/blob/master/Libraries/Components/Touchable/TouchableHighlight.js#L382) to attach that ref to the native host ref. That means we can do this:

```
<TouchableHighlight ref={ref => {ref.focus()}/>
```
With this component it seems like you need to do

```
<TouchableWin32Highlight innerRef={ref => {ref.focus()} />
```

Do you need to have a different prop name or can you use forwardRef to have the same API?

**Comment by tkamireh** [comment]
> In core when you attach a ref to TouchableHighlight, we [use forwardRef](https://github.com/facebook/react-native/blob/master/Libraries/Components/Touchable/TouchableHighlight.js#L382) to attach that ref to the native host ref. That means we can do this:
> 
> ```
> <TouchableHighlight ref={ref => {ref.focus()}/>
> ```
> 
> With this component it seems like you need to do
> 
> ```
> <TouchableWin32Highlight innerRef={ref => {ref.focus()} />
> ```
> 
> Do you need to have a different prop name or can you use forwardRef to have the same API?

Hey, nice to see people from core here 😃! I suppose we could use forwardRef and avoid the innerRef prop name, it should be possible to get proper typings. Worth also noting, this is just a test/example component, not meant for use outside of our testing.

While we have you here @TheSavior, do you have thoughts around intended [setAndForwardRef](https://github.com/facebook/react-native/blob/91f139b94118fe8db29728ea8ad855fc4a13f743/Libraries/Utilities/setAndForwardRef.js) usage? I see people use it in some strange ways to do things like ref mutation (e.g. adding functions to the forwarded ref). It seems like everyone would be better off just providing an imperative interface that wraps calls to an internal ref in that case, when they need to do something like that.

**Comment by TheSavior** [comment]
I think setAndForwardRef is just a hack for class components. We are moving everything to functions with forwardRef, so it is just temporary. If I recall correctly, the reason we needed that and not just proxying all the methods is that native refs have other properties on them. For many of the fabric compatible apis, it requires passing a native ref because it uses those other properties that are private. Things like view commands and ref.measureLayout.

**Comment by TheSavior** [comment]
When using forwardRef it is really easy to break display names. We have a test helper we use to make sure that things work as expected and have the right display names. We try to have a jest test for each component that does this to catch regressions. You can see an example for TextInput here. https://github.com/facebook/react-native/blob/7e343c8d3a6865a6ec76449e7159133426743660/Libraries/Components/TextInput/__tests__/TextInput-test.js#L169-L177

with the resulting snapshots here: https://github.com/facebook/react-native/blob/7e343c8d3a6865a6ec76449e7159133426743660/Libraries/Components/TextInput/__tests__/__snapshots__/TextInput-test.js.snap#L3-L73

**Comment by tkamireh** [comment]
w.r.t forwardRef breaking displayNames, are you referring to ref forwarding chains generating long display names (using the default React behavior). Setting displayName on the component returned by forwardRef is always an option here, right? Admittedly, we've haven't done a great job being diligent around displayNames and devtool usage in general, in the win32 space. Snapshot testing is something we have high up on our backlog - I'll be sure to keep this in mind.

**Comment by TheSavior** [comment]
I think sometimes you have to help the default React behavior. If it’s right it will be something like ForwardRef(TextInput). If it is wrong it will be something like ForwardRef(Component)

**Pull request merged**